cloud_sql/postgres/pdo/src/app.php

```diff
@@ -0,0 +1,116 @@
+<?php
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+declare(strict_types=1);
+
+use Google\Cloud\Samples\CloudSQL\Postgres\Votes;
+use Pimple\Container;
+use Pimple\Psr11\Container as Psr11Container;
+use Slim\Factory\AppFactory;
+use Slim\Views\Twig;
+use Slim\Views\TwigMiddleware;
+
+// Create and set the dependency injection container.
+$container = new Container;
+AppFactory::setContainer(new Psr11Container($container));
+
+// add the votes manager to the container.
+$container['votes'] = function (Container $container) {
+    return new Votes($container['db']);
+};
+
+// Setup the database connection in the container.
+$container['db'] = function () {
+    $username = getenv('DB_USER');
+    $password = getenv('DB_PASS');
+    $dbName = getenv('DB_NAME');
+    $hostname = getenv('DB_HOSTNAME');
+    $cloud_sql_connection_name = getenv('CLOUD_SQL_CONNECTION_NAME');
+
+    try {
+        // # [START cloud_sql_postgres_pdo_create]
+        // // $username = 'your_db_user';
+        // // $password = 'yoursupersecretpassword';
+        // // $dbName = 'your_db_name';
+        // // $cloud_sql_connection_name = getenv("CLOUD_SQL_CONNECTION_NAME");
+        // // $hostname = "127.0.0.1"; // Only used in TCP mode.
+
+        if ($hostname) {
+            // Connect using TCP
+            $dsn = sprintf('pgsql:dbname=%s;host=%s', $dbName, $hostname);
+        } else {
+            // Connect using UNIX sockets
+            $dsn = sprintf(
+                'pgsql:dbname=%s;host=/cloudsql/%s',
+                $dbName,
+                $cloud_sql_connection_name
+            );
+        }
+
+        // Connect to the database.
+        // Here we set the connection timeout to five seconds and ask PDO to
+        // throw an exception if any errors occur.
+        $conn = new PDO($dsn, $username, $password, [
+            PDO::ATTR_TIMEOUT => 5,
+            PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION
```

Comment:
Question: Would it make sense to move lines 65 and 66 to after 67, and wrap that section in the `cloud_sql_postgres_pdo_timeout` region tag? For reference, the snippet would be included here: https://cloud.google.com/sql/docs/postgres/manage-connections#timeout
Is this functionality similar to the samples for other languages in this section?

```suggestion
        $conn = new PDO($dsn, $username, $password, [
            // [START cloud_sql_postgres_pdo_timeout]
            // Here we set the connection timeout to five seconds and ask PDO to
            // throw an exception if any errors occur.
            PDO::ATTR_TIMEOUT => 5,
            PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION
            // [END cloud_sql_postgres_pdo_timeout]
```

Reply:
That block is already a part of the `cloud_sql_postgres_pdo_create` region. Is it possible to have nested regions?

Elsewhere I think we wanted to have the timeout and error mode as part of the other snippet, but those settings can also be set separately and duplicating them won't cause any trouble. We could do something like this:

```php
        // # [START cloud_sql_postgres_pdo_create]
        // // $username = 'your_db_user';
        // // $password = 'yoursupersecretpassword';
        // // $dbName = 'your_db_name';
        // // $cloud_sql_connection_name = getenv("CLOUD_SQL_CONNECTION_NAME");
        // // $hostname = "127.0.0.1"; // Only used in TCP mode.

        if ($hostname) {
            // Connect using TCP
            $dsn = sprintf('pgsql:dbname=%s;host=%s', $dbName, $hostname);
        } else {
            // Connect using UNIX sockets
            $dsn = sprintf(
                'pgsql:dbname=%s;host=/cloudsql/%s',
                $dbName,
                $cloud_sql_connection_name
            );
        }

        // Connect to the database.
        // Here we set the connection timeout to five seconds and ask PDO to
        // throw an exception if any errors occur.
        $conn = new PDO($dsn, $username, $password, [
            PDO::ATTR_TIMEOUT => 5,
            PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION
        ]);
        # [END cloud_sql_postgres_pdo_create]

        # [START cloud_sql_postgres_pdo_timeout]
        $conn->setAttribute(PDO::ATTR_TIMEOUT, 5);
        $conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
        # [END cloud_sql_postgres_pdo_timeout]
```

It's a bit odd in the application to set those values twice, but for the purpose of the docs site it may be alright.

Reply:
It is possible to nest region tags, so we won't need to duplicate them. In other samples, we've been using a combination of nesting and using `EXCLUDE` tags, but I think I actually like the approach of setting them outside of the connection creation statement better (as opposed to setting them twice). It looks cleaner. 
```
// # [START cloud_sql_postgres_pdo_create]
// // $username = 'your_db_user';
// // $password = 'yoursupersecretpassword';
// // $dbName = 'your_db_name';
// // $cloud_sql_connection_name = getenv("CLOUD_SQL_CONNECTION_NAME");
// // $hostname = "127.0.0.1"; // Only used in TCP mode.

if ($hostname) {
    // Connect using TCP
    $dsn = sprintf('pgsql:dbname=%s;host=%s', $dbName, $hostname);
} else {
    // Connect using UNIX sockets
    $dsn = sprintf(
        'pgsql:dbname=%s;host=/cloudsql/%s',
        $dbName,
        $cloud_sql_connection_name
    );
}

// Connect to the database.
$conn = new PDO($dsn, $username, $password);
# [END cloud_sql_postgres_pdo_create]

# [START cloud_sql_postgres_pdo_timeout]
// Here we set the connection timeout to five seconds and ask PDO to
// throw an exception if any errors occur.
$conn->setAttribute(PDO::ATTR_TIMEOUT, 5);
$conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
# [END cloud_sql_postgres_pdo_timeout]
```

Reply:
@kurtisvg thoughts?

Reply:
Does `new PDO(...)` initialize the connection on object init, or is it waiting until it's used? If the first, I think we have to include it as part of the initial call (otherwise it won't actually timeout).

If so, I think something like this:

```php
// [START cloud_sql_postgres_pdo_create]
// ...

// Connect to the database.
# [START cloud_sql_postgres_pdo_timeout]
// Here we set the connection timeout to five seconds and ask PDO to
// throw an exception if any errors occur.
$conn = new PDO($dsn, $username, $password, [
    PDO::ATTR_TIMEOUT => 5,
    PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION
]);
# [END cloud_sql_postgres_pdo_timeout]
# [END cloud_sql_postgres_pdo_create]
```

Reply:
it initializes the connection when the object is created.

Reply:
In that case, let's nest the region tags. We should probably update the mysql and sql server samples too. Feel free to open a separate PR for that and I can approve and merge those as well.

Reply:
Hmm. Are we thinking something like this?

```php
// # [START cloud_sql_postgres_pdo_create]
// // $username = 'your_db_user';
// // $password = 'yoursupersecretpassword';
// // $dbName = 'your_db_name';
// // $cloud_sql_connection_name = getenv("CLOUD_SQL_CONNECTION_NAME");
// // $hostname = "127.0.0.1"; // Only used in TCP mode.

if ($hostname) {
    // Connect using TCP
    $dsn = sprintf('pgsql:dbname=%s;host=%s', $dbName, $hostname);
} else {
    // Connect using UNIX sockets
    $dsn = sprintf(
        'pgsql:dbname=%s;host=/cloudsql/%s',
        $dbName,
        $cloud_sql_connection_name
    );
}

// Connect to the database.
// Here we set the connection timeout to five seconds and ask PDO to
// throw an exception if any errors occur.
$conn = new PDO($dsn, $username, $password, [
    PDO::ATTR_TIMEOUT => 5,
    PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION
]);

# [START_EXCLUDE]
# [START cloud_sql_postgres_pdo_timeout]
$conn->setAttribute(PDO::ATTR_TIMEOUT, 5);
$conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
# [END cloud_sql_postgres_pdo_timeout]
# [END_EXCLUDE]
# [END cloud_sql_postgres_pdo_create]
```

Reply:
That won't work because we are still showing our users the wrong way to do it. Was there something wrong with my example above?

The snippet lookup automatically pulls out comments that are region tags, so it'll render without the extra region tags (but the code will still be included.

For example:
```
# [START REGION1]
This shows in region 1. 
# [START REGION2]
This shows in region 1 and 2. 
# [END REGION2]
# [END REGION1]
```

Reply:
Ah, sorry I misunderstood at first. This is fixed now.